Filter member email recipients based on the newsletter subscriptions

tpatel opened the pull request:
refs https://github.com/TryGhost/Team/issues/1524

- We need to fetch the post newsletter to grab the slug as it's needed for the member NQL filter.
- We can then use the newsletter slug and append it in the existing member NQL filter.

This is the backend implementation only, there is still the need for an admin PR to update the members counts that are displayed in the Publish flow.

tpatel committed:
Filter member email recipients based on the newsletter subscriptions

refs https://github.com/TryGhost/Team/issues/1524

- We need to fetch the post newsletter to grab the slug as it's needed for the member NQL filter.
- We can then use the newsletter slug and append it in the existing member NQL filter.

matthanley commented on core/server/services/mega/mega.js:
```diff
@@ -384,6 +385,11 @@ async function getEmailMemberRows({emailModel, memberSegment, options}) {
 
+    if (labsService.isSet('multipleNewsletters') && emailModel.get('newsletter_id')) {
+        const newsletter = await emailModel.related('newsletter').fetch();
+        filterOptions.filter = `${filterOptions.filter}+newsletters:${newsletter.slug}`;
```
While it's not used right now, I think we should also take into account the newsletter's `visibility` property here to satisfy [this requirement](https://www.notion.so/ghost/Architecture-Notes-65c8733988e946d98872bb9f4655b40f#e656ac2e3d974b2eaecc911812f0173a).

matthanley commented:
@tpatel you'll also need to update [the transform](https://github.com/tpatel/Ghost/blob/211497ce95c54b3c6cad65e51178f488cff3c10f/core/server/services/mega/mega.js#L130) to stop using the member's subscribed flag.

tpatel committed:
Updated transformEmailRecipientFilter to fully support filtering members

- Removed `subscribed:true` when an email is sent to a newsletter
- Added `status:-free` when an email is sent to a newsletter with `visibility` set to `paid`

tpatel committed:
fixed a typo, added a test, checking the coverage % reported by the CI

kevinansfield commented on core/server/services/mega/mega.js:
```diff
@@ -377,7 +413,15 @@ async function getEmailMemberRows({emailModel, memberSegment, options}) {
-    const recipientFilter = transformEmailRecipientFilter(emailModel.get('recipient_filter'), {errorProperty: 'recipient_filter'});
+    let newsletter;
+    if (labsService.isSet('multipleNewsletters')) {
+        try {
+            newsletter = await emailModel.related('newsletter').fetch();
+        } catch (error) {
+            // Ignore NotFoundError
+        }
```
Can you use `const newsletter = await emailModel.related('newsletter').fetch({require: false})` instead here? My understanding is that would [ignore a not-found error](https://bookshelfjs.org/api.html#Model-instance-fetch) automatically, in which case if there's some other error we won't be silently hiding it.

tpatel replied:
neat, I didn't know about this property, thanks 👍

tpatel committed:
Added transaction and require options to the newsletter fetch

Commit:
Added newsletter id filter to transformEmailRecipientFilter

refs https://github.com/TryGhost/Ghost/pull/14489

- The member recipients were not filtered by the newsletter they should be subscribed to
- Removed duplicate labs import
- Requires fixtures changes to update the tests

SimonBackx commented:
I'm working on some changes here with the fixtures from https://github.com/TryGhost/Ghost/pull/14563. Still some failing tests but now they should be easy to solve with these fixtures

Commit:
Updated posts regression tests for new newsletter fixtures

refs https://github.com/TryGhost/Ghost/pull/14489

- Fixtures from https://github.com/TryGhost/Ghost/pull/14563 (not yet in this commit)
- Added tests what happens when you publish without newsletter_id
- Added tests what happens when you publish with newsletter_id
- The tests without newsletter_id are currently failing

SimonBackx commented:
There is still a bug somewhere when sending without the newsletter_id option

Commit:
Fixed default newsletter filtering and tests

refs https://github.com/TryGhost/Ghost/pull/14489

The status filtering of newsletters used the wrong format. Updated the
posts regression tests to make sure the default newsletter fixture is
archived before starting with the tests.

Commit:
Updated mega tests

refs https://github.com/TryGhost/Ghost/pull/14489

- transformEmailRecipientFilter test updated to include newsletter id
- getEmailMemberRows: added mocked related relation

Commit:
Removed posts service debug query parameters

no issue

Removed query select and limit parameters used for debugging.

Commit:
Updated posts test check whether newsletter_id can't be changed

no issue

The newsletter_id can be changed if no email was sent. Since no email service
is running in the tests, the newsletter_id was allowed to change.
